packages/pyright-internal/src/analyzer/sourceFile.ts

```diff
@@ -536,23 +552,59 @@ export class SourceFile {
     }
 
     getImports(): ImportResult[] {
-        return this._imports || [];
+        return this._writableData.imports || [];
     }
 
     getBuiltinsImport(): ImportResult | undefined {
-        return this._builtinsImport;
+        return this._writableData.builtinsImport;
     }
 
     getIPythonDisplayImport(): ImportResult | undefined {
-        return this._ipythonDisplayImport;
+        return this._writableData.ipythonDisplayImport;
     }
 
     getModuleSymbolTable(): SymbolTable | undefined {
-        return this._moduleSymbolTable;
+        return this._writableData.moduleSymbolTable;
     }
 
     getCheckTime() {
-        return this._checkTime;
+        return this._writableData.checkTime;
+    }
+
+    enterEditMode() {
+        this._isEditMode = true;
+    }
+
+    exitEditMode(): FileEditAction | undefined {
```

Comment:
I was hoping to remove knowledge of VSCode's FileEditAction and TextDocument from this file. Could we simply return the new contents as a string and let the caller construct the FileEditAction from it?

Reply:
Sure that works. You mean at the program level?

Reply:
Actually, I should have said "languageServerBase", not "service".

Reply:
Something needs the file path though? You mean create another interface that just stores the content and the file path?

Reply:
Ah, I see what you mean. Yeah, then keeping it in the service (in the `useEditMode` method) makes sense.